File: z3r0sphere-ssd-logic/spherical_kernel.py

def l2_norm(v):
    """Compute the Euclidean (L2) norm of a vector."""
    total = 0.0
    for x in v:
        total += x * x
    return total ** 0.5

def dot_product(a, b):
    """Compute the dot product of two vectors."""
    return sum(x * y for x, y in zip(a, b))

def spherical_kernel(x, y):
    """Cosine similarity kernel (unit-sphere projection, pure Python, no dependencies)."""
    x_norm = l2_norm(x)
    y_norm = l2_norm(y)
    if x_norm == 0.0 or y_norm == 0.0:
        raise ValueError("Zero vector not allowed.")
    return dot_product(x, y) / (x_norm * y_norm)

if __name__ == "__main__":
    # Example vectors
    examples = [
        ([1, 0, 0], [0, 1, 0]),
        ([1, 2, 3], [4, 5, 6]),
        ([1, 0, 0], [1, 0, 0]),
        ([1, 0, 0], [-1, 0, 0]),
    ]
    for x, y in examples:
        print(f"x = {x}, y = {y}, cosine similarity = {spherical_kernel(x, y)}")

File: z3r0sphere-ssd-logic/spherical_kernel.c
#include <math.h>
#include <stdio.h>

float spherical_kernel(const float *x, const float *y, int d) {
    float x_norm = 0.0f, y_norm = 0.0f, dot = 0.0f;
    for (int i = 0; i < d; ++i) {
        x_norm += x[i] * x[i];
        y_norm += y[i] * y[i];
        dot    += x[i] * y[i];
    }
    x_norm = sqrtf(x_norm);
    y_norm = sqrtf(y_norm);
    if (x_norm == 0.0f || y_norm == 0.0f) return 0.0f; // or handle error
    return dot / (x_norm * y_norm);
}

int main() {
    float x[3] = {1, 2, 3};
    float y[3] = {4, 5, 6};
    printf("Cosine similarity: %f\n", spherical_kernel(x, y, 3));
    return 0;
}
